fendermint/vm/topdown/proof-service/src/cache.rs

```diff
@@ -0,0 +1,418 @@
+// Copyright 2022-2025 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! In-memory cache for proof bundles with optional disk persistence
+
+use crate::config::CacheConfig;
+use crate::observe::{ProofCached, CACHE_HIT_TOTAL, CACHE_SIZE};
+use crate::persistence::ProofCachePersistence;
+use crate::types::CacheEntry;
+use anyhow::{Context, Result};
+use ipc_observability::emit;
+use parking_lot::RwLock;
+use std::collections::BTreeMap;
+use std::path::Path;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+/// Thread-safe in-memory cache for proof bundles
+#[derive(Clone)]
+pub struct ProofCache {
+    /// Map: instance_id -> CacheEntry
+    /// Using BTreeMap for ordered iteration
+    entries: Arc<RwLock<BTreeMap<u64, CacheEntry>>>,
+
+    /// Last committed instance ID (updated after execution)
+    last_committed_instance: Arc<AtomicU64>,
+
+    /// Configuration
+    config: CacheConfig,
+
+    /// Optional disk persistence
+    persistence: Option<Arc<ProofCachePersistence>>,
+}
+
+impl ProofCache {
+    /// Create a new proof cache (in-memory only)
+    pub fn new(last_committed_instance: u64, config: CacheConfig) -> Self {
+        Self {
+            entries: Arc::new(RwLock::new(BTreeMap::new())),
+            last_committed_instance: Arc::new(AtomicU64::new(last_committed_instance)),
+            config,
+            persistence: None,
+        }
+    }
+
+    /// Create a new proof cache with disk persistence
+    ///
+    /// Loads existing entries from disk on startup.
+    /// If DB is fresh, uses `initial_instance` as the starting point.
+    pub fn new_with_persistence(
+        config: CacheConfig,
+        db_path: &Path,
+        initial_instance: u64,
+    ) -> Result<Self> {
+        let persistence = ProofCachePersistence::open(db_path)?;
+
+        // Load all entries and last committed instance from disk
+        let last_committed = persistence
+            .load_last_committed()
+            .context("Failed to load last committed instance from disk")?
+            .unwrap_or(initial_instance);
```

Comment:
What's the point of storing the last committed instance ID in the cache's persistent storage? Just for debugging when we'd inspect the storage? (We anyways cannot guarantee that it always matches the chain state.)